### DISCUSSION TITLE
Help creating an outline from an image

### DISCUSSION THREAD

**Discussion by techjp** [ImageMagick/ImageMagick#6431] [Help] [answered]
### ImageMagick version

7.1.1-11

### Operating system, version and so on

Windows 10, fully updated

### Description

I am trying to take a source image of Japanese calligraphy characters and create outlines of them. A few details about these images:

1. They are PNG files.
2. They have transparent backgrounds.
3. The images are very clean, so it is easy to select just the non-transparent parts.
4. The selection must include all non-transparent pixels, meaning a non-contiguous selection.
5. The output, once the outline has been created, should be transparent where there is no data.

 In Photoshop I can do this via the following steps:

1. Use the Magic Wand tool to select all transparent parts of the image. "Contiguous" is deselected so ALL transparent areas get selected with one click.
2. Invert the selection so now instead of all transparent areas being selected, now the character (ie non-transparent) area is selected.
3. Contract the selection by 15 pixels.
4. Delete all the still-selected pixels.

This results in a good enough outline of the source character(s) for my needs.

As a sample image, (though in JPG with no transparency, unfortunately), I start with this image:
![image](https://github.com/ImageMagick/ImageMagick/assets/7457115/59558f83-457c-4e02-a5b2-084a242fb8d2)

And after following the above steps in Photoshop, I get this image:
![image](https://github.com/ImageMagick/ImageMagick/assets/7457115/b2e26226-0947-43cd-b64b-04401827047c)

I have spent hours trying to get this to work in ImageMagick but have failed. I've created masks, shrunk the masks, then tried to apply them, but I do not get an outline. I'm sure I'm missing something simple.

Help would be greatly appreciated.

**Comment by GeeMack** [comment, marked as answer]
Just a quick test using ImageMagick's "-morphology" operator with your sample image as a black and white, I get the black on transparent output below....

    magick bandw.png -background none -morphology edge octagon -negate -fuzz 33% -transparent white result.png


![result](https://github.com/ImageMagick/ImageMagick/assets/65475279/ed5ebc02-20b6-41bc-9933-308892e82486)

Edited to add: When I start with your sample as a black-on-transparent image, this command nearly achieves your example result...

    magick transparent_bg.png -background none -morphology edgein octagon:3 result2.png

That command produces this...

![result2](https://github.com/ImageMagick/ImageMagick/assets/65475279/6a614ffc-8f53-4018-8dcb-0f7676014358)

The the thickness of the lines in the result can be adjusted by various options to the "-morphology" operator. See this page for more information...

[https://legacy.imagemagick.org/Usage/morphology/](url)

**Comment by techjp** [comment]
Thank you! That works quite well, but it leaves some light gray content inside the original image. It's hard to see on the screen with a transparent background but it shows up on a white background or when printing. So I'm trying to find a way to remove that. I think if you zoom in on this image that you will be able to see it:

![image](https://github.com/ImageMagick/ImageMagick/assets/7457115/18492f5a-3aff-4ce5-82c7-f085aba03597)

Is there a way to adjust the morphology to have a cleaner outline? I also looked around for a command that might clean up a lot of it and came across this:
https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=36051
with this command:
`convert in.jpg \( +clone -colorspace HCL -channel G -separate +channel -threshold 15% \) -alpha off -compose CopyOpacity -composite out.png`

The command looks promising but for the life of me I can't convert that Linux command line to work on Windows. I'm not sure what those brackets are trying to accomplish but whatever it is, I can't duplicate it!

Thank you again for the morphology operator!

**Reply by fmw42** [reply]
The parentheses are needed as they do processing on the side so that one does not process the input. I believe the Windows version would be (remove the \ in front of the parentheses).

`convert in.jpg ( +clone -colorspace HCL -channel G -separate +channel -threshold 15% ) -alpha off -compose CopyOpacity -composite out.png`

If in a bat file, double the % to %%

**Reply by techjp** [reply]
Thanks for the reply.

I definitely tried removing the escape characters but I got invalid parameter responses. It turned out to be that `convert` is no longer supported, must be `magick convert` instead. So the command itself works now but all I get as output is an entirely black box. :(

Have some more work to do to figure this out, I think.

**Comment by snibgo** [comment]
> I also looked around for a command that might clean up a lot of it and came across this: ...

But that command doesn't do what you want. That command is for "Removing [making transparent] All Shades of Grey from an Image". In that context, black and white are shades of gray.

What you want to do (perhaps) is to convert all pixels that are neither black nor white into either black or white. This is a simple threshold:
```
magick in.png -threshold 50% out.png
```
If in a bat file, double the `%` to `%%`.

**Reply by techjp** [reply]
That does indeed remove all the internal gray area, but it also removes all the smoothness gradient around the outline of the character as well, resulting in a line with really jagged edges.

I'm wondering if there is any way to combine that with a mask that would protect the edges I want to keep smooth while removing all the gray areas inside the outline?

**Reply by GeeMack** [reply]
You can use "-evaluate" to tweak the result just before the command output so all pixels are either black or transparent while keeping the alpha aliasing. 

    ...
    -channel RGB -evaluate set 0 +channel ^
    ...

That will isolate the RGB channels and take all color out of them, leaving the alpha channel alone. That should get rid of your little grayish marks.

Edited to add a couple things to try:

If your working images are only the size of your example, you might consider super-sampling. Essentially you resize the input image 200%, do all the operations on it, then resize it 50% to the original dimensions. That can make the edges smoother.

Possibly try using the "-morphology" operation to create a much fatter outline than you want, then use the "erode" option to "-morphology" to make another pass and peel off a tiny bit. That might help get more even, smoother edged lines.

**Comment by snibgo** [comment]
> I'm wondering if there is any way to combine that with a mask that would protect the edges I want to keep smooth while removing all the gray areas inside the outline?

Yes, if you can define that mask. You could draw it manually, of course. To create it automatically, we need to know what distinguishes "the edges I want to keep smooth" from "all the gray areas inside the outline".

**Comment by techjp** [comment]
Just to give everyone some closure (and for any future people who come across this topic!), I was able to solve this using a combination of imagemagick and potrace. I used imagemagick to convert to a PGM file with a white background. Then I was able to use potrace to produce an SVG file. potrace does not provide just an outline but I was able to edit the SVG file (can automate that) to make the fill white and the outline black. Worked out well!
